python/ray/tests/test_memory_deadlock.py

```diff
@@ -0,0 +1,198 @@
+    for _ in range(chunks):
+        mem.append([0] * bytes_per_chunk)
+    return mem
+
+
+@pytest.mark.skipif(
+    sys.platform != "linux" and sys.platform != "linux2",
+    reason="memory monitor only on linux currently",
+)
+def test_churn_long_running(
+    ray_with_memory_monitor,
+):
+    long_running_bytes = get_additional_bytes_to_reach_memory_usage_pct(
+        memory_usage_threshold_fraction - 0.1
+    )
+    ray.get(
+        allocate_memory.options(max_retries=1).remote(
+            long_running_bytes, post_allocate_sleep_s=30
+        )
+    )
+    small_bytes = get_additional_bytes_to_reach_memory_usage_pct(
+        memory_usage_threshold_fraction + 0.2
+    )
+    with pytest.raises(ray.exceptions.OutOfMemoryError) as _:
+        ray.get(allocate_memory.options(max_retries=1).remote(small_bytes))
+
+
+@pytest.mark.skipif(
+    sys.platform != "linux" and sys.platform != "linux2",
+    reason="memory monitor only on linux currently",
+)
+def test_deadlock_task_with_nested_task(
+    ray_with_memory_monitor,
+):
+    with pytest.raises(ray.exceptions.OutOfMemoryError) as _:
+        bytes1 = get_additional_bytes_to_reach_memory_usage_pct(
+            memory_usage_threshold_fraction - 0.1
+        )
+        bytes2 = (
+            get_additional_bytes_to_reach_memory_usage_pct(
+                memory_usage_threshold_fraction + 0.2
+            )
+            - bytes1
+        )
+        ray.get(
+            task_with_nested_task.remote(
+                task_bytes=bytes1, nested_task_bytes=bytes2, barrier=None
+            )
+        )
+
+
+@ray.remote
+def task_with_nested_actor(
+    first_fraction, second_fraction, actor_allocation_first=True
+):
+    leaker = Leaker.options(max_restarts=1, max_task_retries=1).remote()
+    first_bytes = get_additional_bytes_to_reach_memory_usage_pct(first_fraction)
+    if actor_allocation_first:
+        ray.get(leaker.allocate.remote(first_bytes))
+        second_bytes = get_additional_bytes_to_reach_memory_usage_pct(second_fraction)
+        dummy = alloc_mem(second_bytes)
+    else:
+        dummy = alloc_mem(first_bytes)
+        second_bytes = get_additional_bytes_to_reach_memory_usage_pct(second_fraction)
+        ray.get(leaker.allocate.remote(second_bytes))
+
+
+@pytest.mark.skipif(
+    sys.platform != "linux" and sys.platform != "linux2",
+    reason="memory monitor only on linux currently",
+)
+def test_deadlock_task_with_nested_actor_swap_order(
+    ray_with_memory_monitor,
+):
+    with pytest.raises(ray.exceptions.RayTaskError) as _:
+        ray.get(
+            task_with_nested_actor.remote(
+                first_fraction=memory_usage_threshold_fraction - 0.1,
+                second_fraction=memory_usage_threshold_fraction + 0.25,
+                actor_allocation_first=False,
+            )
+        )
+
+
+@ray.remote
+class ActorWithNestedTask:
+    def __init__(self):
+        self.mem = []
+
+    def alloc_local(self, num_bytes):
```

Comment:
this isn't quite correct, it should be a single remote call that invokes a nested task

instead of making 2 calls it should use a barrier object to synchronize